File: 0x4/bsearch_test.c
/* SDPX-License-Identifier: GPL-2.0 */

#include <stdio.h>

#include "bsearch.h"

int test_bsearch0(int *i)
{
	const struct test {
		const char	*name;
		const int	array[1];
		int		value;
		int		want;
	} tests[] = {
		{
			.name	= "no entry",
			.array	= {},
			.value	= 10,
			.want	= -1,
		},
		{ /* sentry */ },
	};
	const struct test *t;
	int fail = 0;

	for (t = tests; t->name; t++) {
		int got;

		printf("%2d) %-15s: %-55s", ++(*i), "bsearch0_test", t->name);
		got = bsearch(t->value, t->array, 0);
		if (got != t->want) {
			printf("FAIL: %d!=%d\n", got, t->want);
			fail++;
		} else
			puts("PASS");
	}
	return fail;
}

int test_bsearch2(int *i)
{
	const struct test {
		const char	*name;
		const int	array[2];
		int		value;
		int		want;
	} tests[] = {
		{
			.name	= "first entry",
			.array	= {0, 1},
			.value	= 0,
			.want	= 0,
		},
		{
			.name	= "last entry",
			.array	= {0, 1},
			.value	= 1,
			.want	= 1,
		},
		{
			.name	= "no entry",
			.array	= {0, 1},
			.value	= 10,
			.want	= -1,
		},
		{
			.name	= "first entry",
			.array	= {0, 2},
			.value	= 0,
			.want	= 0,
		},
		{
			.name	= "last entry",
			.array	= {0, 2},
			.value	= 2,
			.want	= 1,
		},
		{
			.name	= "no entry",
			.array	= {0, 2},
			.value	= 20,
			.want	= -1,
		},
		{ /* sentry */ },
	};
	const struct test *t;
	int fail = 0;

	for (t = tests; t->name; t++) {
		int got;

		printf("%2d) %-15s: %-55s", ++(*i), "bsearch2_test", t->name);
		got = bsearch(t->value, t->array, 2);
		if (got != t->want) {
			printf("FAIL: %d!=%d\n", got, t->want);
			fail++;
		} else
			puts("PASS");
	}
	return fail;
}

int test_bsearch3(int *i)
{
	const struct test {
		const char	*name;
		const int	array[3];
		int		value;
		int		want;
	} tests[] = {
		{
			.name	= "first entry",
			.array	= {0, 1, 2},
			.value	= 0,
			.want	= 0,
		},
		{
			.name	= "last entry",
			.array	= {0, 1, 2},
			.value	= 2,
			.want	= 2,
		},
		{
			.name	= "middle entry",
			.array	= {0, 1, 2},
			.value	= 1,
			.want	= 1,
		},
		{
			.name	= "no entry",
			.array	= {0, 1, 2},
			.value	= 10,
			.want	= -1,
		},
		{
			.name	= "first entry",
			.array	= {0, 2, 4},
			.value	= 0,
			.want	= 0,
		},
		{
			.name	= "last entry",
			.array	= {0, 2, 4},
			.value	= 4,
			.want	= 2,
		},
		{
			.name	= "middle entry",
			.array	= {0, 2, 4},
			.value	= 2,
			.want	= 1,
		},
		{
			.name	= "no entry",
			.array	= {0, 2, 4},
			.value	= 20,
			.want	= -1,
		},
		{ /* sentry */ },
	};
	const struct test *t;
	int fail = 0;

	for (t = tests; t->name; t++) {
		int got;

		printf("%2d) %-15s: %-55s", ++(*i), "bsearch3_test", t->name);
		got = bsearch(t->value, t->array, 3);
		if (got != t->want) {
			printf("FAIL: %d!=%d\n", got, t->want);
			fail++;
		} else
			puts("PASS");
	}
	return fail;
}

int test_bsearch10(int *i)
{
	const struct test {
		const char	*name;
		const int	array[10];
		int		value;
		int		want;
	} tests[] = {
		{
			.name	= "first entry",
			.array	= {0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
			.value	= 0,
			.want	= 0,
		},
		{
			.name	= "last entry",
			.array	= {0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
			.value	= 9,
			.want	= 9,
		},
		{
			.name	= "middle entry",
			.array	= {0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
			.value	= 5,
			.want	= 5,
		},
		{
			.name	= "no entry",
			.array	= {0, 1, 2, 3, 4, 5, 6, 7, 8, 9},
			.value	= 10,
			.want	= -1,
		},
		{
			.name	= "first entry",
			.array	= {0, 2, 4, 6, 8, 10, 12, 14, 16, 18},
			.value	= 0,
			.want	= 0,
		},
		{
			.name	= "last entry",
			.array	= {0, 2, 4, 6, 8, 10, 12, 14, 16, 18},
			.value	= 18,
			.want	= 9,
		},
		{
			.name	= "middle entry",
			.array	= {0, 2, 4, 6, 8, 10, 12, 14, 16, 18},
			.value	= 10,
			.want	= 5,
		},
		{
			.name	= "no entry",
			.array	= {0, 2, 4, 6, 8, 10, 12, 14, 16, 18},
			.value	= 20,
			.want	= -1,
		},
		{ /* sentry */ },
	};
	const struct test *t;
	int fail = 0;

	for (t = tests; t->name; t++) {
		int got;

		printf("%2d) %-15s: %-55s", ++(*i), "bsearch10_test", t->name);
		got = bsearch(t->value, t->array, 10);
		if (got != t->want) {
			printf("FAIL: %d!=%d\n", got, t->want);
			fail++;
		} else
			puts("PASS");
	}
	return fail;
}

File: 0x4/bsearch/bsearch_test.go
// SPDX-License-Identifier: GPL-2.0
package bsearch

import "testing"

func TestBSearchIntOn10(t *testing.T) {
	tests := []struct {
		name   string
		needle int
		want   int
	}{
		{
			name:   "first entry out of 10 elements",
			needle: 0,
			want:   0,
		},
		{
			name:   "last entry out of 10 elements",
			needle: 9,
			want:   9,
		},
		{
			name:   "middle entry out of 10 elements",
			needle: 4,
			want:   4,
		},
	}
	var haystack = []int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}

	for _, tc := range tests {
		tc := tc
		t.Run(tc.name, func(t *testing.T) {
			got := BSearchInt(tc.needle, haystack)
			if got != tc.want {
				t.Fatalf("wrong index:\n- want: %d\n-  got: %d\n",
					tc.want, got)
			}
		})
	}
}

func TestBSearchIntOn1024(t *testing.T) {
	tests := []struct {
		name   string
		needle int
		want   int
	}{
		{
			name:   "first entry out of 1024 elements",
			needle: 0,
			want:   0,
		},
		{
			name:   "last entry out of 1024 elements",
			needle: 1023,
			want:   1023,
		},
		{
			name:   "middle entry out of 1024 elements",
			needle: 511,
			want:   511,
		},
	}
	haystack := make([]int, 1024)
	for i := 0; i < 1024; i++ {
		haystack[i] = i
	}

	for _, tc := range tests {
		tc := tc
		t.Run(tc.name, func(t *testing.T) {
			got := BSearchInt(tc.needle, haystack)
			if got != tc.want {
				t.Fatalf("wrong index:\n- want: %d\n-  got: %d\n",
					tc.want, got)
			}
		})
	}
}

func TestBSearchIntWithTryOn10(t *testing.T) {
	tests := []struct {
		name   string
		needle int
		want   int
		try    int
	}{
		{
			name:   "first entry out of 10 elements",
			needle: 0,
			want:   0,
			try:    3,
		},
		{
			name:   "last entry out of 10 elements",
			needle: 9,
			want:   9,
			try:    4,
		},
		{
			name:   "middle entry out of 10 elements",
			needle: 4,
			want:   4,
			try:    1,
		},
	}
	var haystack = []int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}

	for _, tc := range tests {
		tc := tc
		t.Run(tc.name, func(t *testing.T) {
			got, try := BSearchIntWithTry(tc.needle, haystack)
			if got != tc.want {
				t.Fatalf("wrong index:\n- want: %d\n-  got: %d\n",
					tc.want, got)
			}
			if try != tc.try {
				t.Fatalf("wrong try:\n- want: %d\n-  got: %d\n",
					tc.try, try)
			}
		})
	}
}

func TestBSearchIntWithTryOn1024(t *testing.T) {
	tests := []struct {
		name   string
		needle int
		want   int
		try    int
	}{
		{
			name:   "first entry out of 1024 elements",
			needle: 0,
			want:   0,
			try:    10,
		},
		{
			name:   "last entry out of 1024 elements",
			needle: 1023,
			want:   1023,
			try:    10,
		},
		{
			name:   "middle entry out of 1024 elements",
			needle: 511,
			want:   511,
			try:    1,
		},
	}
	haystack := make([]int, 1024)
	for i := 0; i < 1024; i++ {
		haystack[i] = i
	}

	for _, tc := range tests {
		tc := tc
		t.Run(tc.name, func(t *testing.T) {
			got := BSearchInt(tc.needle, haystack)
			if got != tc.want {
				t.Fatalf("wrong index:\n- want: %d\n-  got: %d\n",
					tc.want, got)
			}
		})
	}
}
